cranelift/filetests/src/test_zkasm.rs

```diff
@@ -0,0 +1,187 @@
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use cranelift_codegen::entity::EntityRef;
+    use cranelift_codegen::ir::function::FunctionParameters;
+    use cranelift_codegen::ir::ExternalName;
+    use cranelift_codegen::isa::zkasm;
+    use cranelift_codegen::{settings, FinalizedMachReloc, FinalizedRelocTarget};
+    use cranelift_wasm::{translate_module, DummyEnvironment};
+
+    fn generate_preamble(start_func_index: usize) -> Vec<String> {
+        let mut program: Vec<String> = Vec::new();
+        program.push("start:".to_string());
+        program.push("  zkPC + 2 => RR".to_string());
+        program.push(format!("  :JMP(function_{})", start_func_index));
+        program.push("  :JMP(finalizeExecution)".to_string());
+        program
+    }
+
+    fn generate_postamble() -> Vec<String> {
+        let mut program: Vec<String> = Vec::new();
+        program.push("finalizeExecution:".to_string());
+        program.push("  ${beforeLast()}  :JMPN(finalizeExecution):".to_string());
+        program.push("                   :JMP(start)".to_string());
+        program
+    }
+
+    fn fix_relocs(
+        code_buffer: &mut Vec<u8>,
+        params: &FunctionParameters,
+        relocs: &[FinalizedMachReloc],
+    ) {
+        let mut delta = 0i32;
+        for reloc in relocs {
+            let start = (reloc.offset as i32 + delta) as usize;
+            let mut pos = start;
+            while code_buffer[pos] != b'\n' {
+                pos += 1;
+                delta -= 1;
+            }
+
+            let code = if let FinalizedRelocTarget::ExternalName(ExternalName::User(name)) =
+                reloc.target
+            {
+                let name = &params.user_named_funcs()[name];
+                if name.index == 0 {
+                    b"  B :ASSERT".to_vec()
+                } else {
+                    format!("  zkPC + 2 => RR\n  :JMP(function_{})", name.index)
+                        .as_bytes()
+                        .to_vec()
+                }
+            } else {
+                b"  UNKNOWN".to_vec()
+            };
+            delta += code.len() as i32;
+
+            code_buffer.splice(start..pos, code);
+        }
+    }
+
+    fn optimize_labels(code: &[&str], func_index: usize) -> Vec<String> {
```

Comment:
Similarly we should probably look into producing the right code from the first principles. It is alright to have this here for the time being, but a `FIXME` would help to make sure this does not get lost.

Reply:
I agree, label optimization should happen in `MachBuffer`, but we don't leverage it yet. Added a TODO